test/csuite/wt3363_checkpoint_op_races/main.c

```diff
@@ -0,0 +1,498 @@
+ * binary, for any purpose, commercial or non-commercial, and by any
+ * means.
+ *
+ * In jurisdictions that recognize copyright laws, the author or authors
+ * of this software dedicate any and all copyright interest in the
+ * software to the public domain. We make this dedication for the benefit
+ * of the public at large and to the detriment of our heirs and
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include "test_util.h"
+
+/*
+ * JIRA ticket reference: WT-3362
+ *
+ * Test case description: There are a number of operations that we run that we
+ * expect not to conflict with or block against a running checkpoint. This test
+ * aims to run repeated checkpoints in a thread, while running an assortment
+ * of operations that we expect to execute quickly on further threads. To
+ * ensure that we catch any blockages we introduce a very large delay into the
+ * checkpoint and measure that no operation takes 1/2 the length of this delay.
+ *
+ * Failure mode: We monitor the execution time of all operations and if we find
+ * any operation taking longer than 1/2 the delay time, we abort dumping a core
+ * file which can be used to determine what operation was blocked.
+ */
+
+int handle_error(WT_EVENT_HANDLER *, WT_SESSION *, int, const char *);
+int handle_message(WT_EVENT_HANDLER *, WT_SESSION *, const char *);
+void* monitor(void *);
+void* do_checkpoints(void *);
+void* do_ops(void *);
+void op_bulk(WT_CONNECTION *);
+void op_bulk_unique(WT_CONNECTION *, int, int);
+void op_create(WT_CONNECTION *);
+void op_create_unique(WT_CONNECTION *, int, int);
+void op_cursor(WT_CONNECTION *);
+void op_drop(WT_CONNECTION *, int);
+
+#define	N_THREADS 10
+#define	RUNTIME 900.0
+
+typedef struct {
+	TEST_OPTS *testopts;
+	int threadnum;
+	int nthread;
+	int done;
+} THREAD_ARGS;
+
+static pthread_rwlock_t single;
+/*
+ * Time delay to introduce into checkpoints in seconds. Should be at-least
+ * double the maximum time that any one of the operations should take. Currently
+ * this is set to 10 seconds and we expect no single operation to take longer
+ * than 5 seconds.
+ */
+static const uint64_t max_execution_time = 10;
+
+/* Int used to make unique table names */
+static uint64_t uid = 1;
+
+/*
+ * An array of the number of operations that each of the worker threads has
+ * performed.
+ */
+static uint64_t thread_counters[N_THREADS];
+static const char *uri;
```

Comment:
I'd expect this to always be accessed via the THREAD_OPTS structure, rather than shared via a global variable. It's not strictly necessary, but global variables are harder to maintain, and this isn't tied to the lifetime of the string which makes me nervous.